airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,445 @@
+
+    def match_request(self, request: Any) -> Optional[AbstractCallRatePolicy]:
+        for matcher, policy in self._policies:
+            if matcher(request):
+                return policy
+        return None
+
+    def acquire_call(self, request: Any, block: bool = True, timeout: Optional[float] = None) -> None:
+        """Try to get a call from budget, will block by default.
+        Matchers will be called sequentially in same the order they were added. The first matcher that returns True will
+
+        :param request:
+        :param block: when true (default) will block the current thread until call credit is available
+        :param timeout: if set will limit maximum time in block, otherwise will wait until credit is available
+        :raises: CallRateLimitHit - when no credits left and if timeout was set the waiting time exceed the timeout
+        """
+
+        policy = self.match_request(request)
+        if policy:
+            self._do_acquire(request=request, policy=policy, block=block, timeout=timeout)
+        elif self._policies:
+            logger.info("no policies matched with requests, allow call by default")
+
+    def update_from_response(self, request: Any, response: Any) -> None:
+        """Update budget information based on response from API
+
+        :param request: the initial request that triggered this response
+        :param response: response from the API
+        """
+        pass
+
+    def _do_acquire(self, request: Any, policy: AbstractCallRatePolicy, block: bool, timeout: Optional[float]) -> None:
+        """Internal method to try to acquire a call credit
+
+        :param request:
+        :param policy:
+        :param block:
+        :param timeout:
+        """
+        last_exception = None
+        # sometimes we spend all budget before a second attempt, so we have few more here
+        for attempt in range(1, self.maximum_attempts_to_acquire):
+            try:
+                policy.try_acquire(request, weight=1)
+                return
+            except CallRateLimitHit as exc:
+                last_exception = exc
+                if block:
+                    if timeout is not None:
+                        time_to_wait = min(timedelta(seconds=timeout), exc.time_to_wait)
+                    else:
+                        time_to_wait = exc.time_to_wait
+
+                    time_to_wait = max(timedelta(0), time_to_wait)  # sometimes we get negative duration
+                    logger.info("reached call limit %s. going to sleep for %s", exc.rate, time_to_wait)
+                    time.sleep(time_to_wait.total_seconds())
+                else:
+                    raise
+
+        if last_exception:
+            logger.info("we used all %s attempts to acquire and failed", self.maximum_attempts_to_acquire)
+            raise last_exception
+
+
+class HttpAPIBudget(APIBudget):
+    """Implementation of AbstractAPIBudget for http"""
+
+    status_codes_for_rate_limit_hit = (429,)
```

Comment:
this is the right default but let's inject this value in the constructor so we don't need to extend the class to support other http codes